File: metabolomics_analysis.R
# Comprehensive Metabolomics Analysis Workflow
# Analysis Date: 2025-09-05
# Data Source: fasting.csv

# Load required libraries
suppressMessages({
  library(readr)
  library(dplyr) 
  library(ggplot2)
  library(pheatmap)
  library(corrplot)
  library(RColorBrewer)
})

cat("=== Comprehensive Metabolomics Analysis Workflow ===\n")
cat("Starting analysis...\n\n")

# Read data with semicolon separator (detected from file inspection)
tryCatch({
  data <- read_csv2("fasting.csv")
  cat("Data loaded successfully with semicolon separator\n")
}, error = function(e) {
  cat("Error reading CSV file:", e$message, "\n")
  cat("Trying comma separator...\n")
  data <<- read_csv("fasting.csv")
})

# Data overview
cat("Data dimensions:", dim(data), "\n")
cat("Sample types found:\n")
print(data[,1])

# Extract numeric data (all columns except first which contains sample IDs)
numeric_data <- data[, -1]
numeric_data <- sapply(numeric_data, as.numeric)
rownames(numeric_data) <- data[[1]]

cat("Numeric data shape:", dim(numeric_data), "\n")

# Basic statistics
summary_stats <- data.frame(
  Metabolite = colnames(numeric_data),
  Mean = colMeans(numeric_data, na.rm = TRUE),
  SD = apply(numeric_data, 2, sd, na.rm = TRUE),
  Min = apply(numeric_data, 2, min, na.rm = TRUE),
  Max = apply(numeric_data, 2, max, na.rm = TRUE),
  Median = apply(numeric_data, 2, median, na.rm = TRUE),
  Q25 = apply(numeric_data, 2, quantile, 0.25, na.rm = TRUE),
  Q75 = apply(numeric_data, 2, quantile, 0.75, na.rm = TRUE)
)

# Save summary statistics
write.csv(summary_stats, "summary_statistics.csv", row.names = FALSE)
cat("Summary statistics saved to summary_statistics.csv\n")

# PCA Analysis
cat("\nPerforming PCA analysis...\n")
# Replace NAs with 0 and scale data
numeric_data_clean <- numeric_data
numeric_data_clean[is.na(numeric_data_clean)] <- 0
pca_result <- prcomp(numeric_data_clean, scale. = TRUE)

# PCA plot
png("pca_analysis.png", width = 1200, height = 900)
par(mfrow = c(2, 2))

# Main PCA plot
sample_colors <- ifelse(grepl("normal", rownames(numeric_data_clean), ignore.case = TRUE), "red", "blue")
plot(pca_result$x[,1], pca_result$x[,2], 
     col = sample_colors, pch = 19, cex = 1.5,
     xlab = paste0("PC1 (", round(summary(pca_result)$importance[2,1]*100, 1), "% variance)"),
     ylab = paste0("PC2 (", round(summary(pca_result)$importance[2,2]*100, 1), "% variance)"),
     main = "PCA Analysis - Sample Distribution")
text(pca_result$x[,1], pca_result$x[,2], rownames(numeric_data_clean), pos = 3, cex = 0.8)
grid(TRUE)
legend("topright", legend = c("Normal", "Fasting"), col = c("red", "blue"), pch = 19)

# Variance explained
var_explained <- summary(pca_result)$importance[2, 1:10]
plot(1:10, cumsum(var_explained), type = "b", pch = 19,
     xlab = "Principal Component", ylab = "Cumulative Variance Explained",
     main = "PCA Variance Explained")
grid(TRUE)

# Top PC1 loadings
pc1_loadings <- sort(abs(pca_result$rotation[,1]), decreasing = TRUE)[1:20]
barplot(pc1_loadings, las = 2, cex.names = 0.7,
        main = "Top 20 PC1 Loadings", ylab = "Absolute Loading")

# Top PC2 loadings  
pc2_loadings <- sort(abs(pca_result$rotation[,2]), decreasing = TRUE)[1:20]
barplot(pc2_loadings, las = 2, cex.names = 0.7,
        main = "Top 20 PC2 Loadings", ylab = "Absolute Loading")

dev.off()
cat("PCA analysis saved to pca_analysis.png\n")

# Correlation Analysis
cat("Performing correlation analysis...\n")
cor_matrix <- cor(numeric_data_clean, use = "complete.obs")

# Find high correlations
high_corr_pairs <- data.frame()
for(i in 1:(ncol(cor_matrix)-1)) {
  for(j in (i+1):ncol(cor_matrix)) {
    corr_val <- cor_matrix[i,j]
    if(abs(corr_val) > 0.7) {
      high_corr_pairs <- rbind(high_corr_pairs, data.frame(
        Metabolite1 = colnames(cor_matrix)[i],
        Metabolite2 = colnames(cor_matrix)[j], 
        Correlation = corr_val
      ))
    }
  }
}
high_corr_pairs <- high_corr_pairs[order(-abs(high_corr_pairs$Correlation)),]
write.csv(high_corr_pairs, "high_correlations.csv", row.names = FALSE)
cat("Found", nrow(high_corr_pairs), "high correlation pairs (>0.7)\n")

# Correlation heatmap (top 50 most variable metabolites)
top_var_metabolites <- names(sort(apply(numeric_data_clean, 2, var), decreasing = TRUE)[1:50])
cor_subset <- cor_matrix[top_var_metabolites, top_var_metabolites]

png("correlation_heatmap.png", width = 1200, height = 1200)
pheatmap(cor_subset, 
         color = colorRampPalette(c("blue", "white", "red"))(100),
         main = "Metabolite Correlation Matrix (Top 50 Most Variable)",
         fontsize = 8)
dev.off()
cat("Correlation heatmap saved to correlation_heatmap.png\n")

# Differential Analysis
cat("Performing differential analysis...\n")
normal_idx <- grep("normal", rownames(numeric_data_clean), ignore.case = TRUE)
fasting_idx <- grep("fasting", rownames(numeric_data_clean), ignore.case = TRUE)

cat("Normal samples:", length(normal_idx), "\n")
cat("Fasting samples:", length(fasting_idx), "\n")

if(length(normal_idx) > 0 & length(fasting_idx) > 0) {
  diff_results <- data.frame()
  
  for(i in 1:ncol(numeric_data_clean)) {
    normal_values <- numeric_data_clean[normal_idx, i]
    fasting_values <- numeric_data_clean[fasting_idx, i]
    
    # T-test
    if(length(normal_values) >= 2 & length(fasting_values) >= 2) {
      ttest <- t.test(normal_values, fasting_values)
      
      normal_mean <- mean(normal_values, na.rm = TRUE)
      fasting_mean <- mean(fasting_values, na.rm = TRUE)
      fold_change <- ifelse(normal_mean > 0, fasting_mean / normal_mean, NA)
      
      diff_results <- rbind(diff_results, data.frame(
        Metabolite = colnames(numeric_data_clean)[i],
        Normal_Mean = normal_mean,
        Fasting_Mean = fasting_mean,
        Fold_Change = fold_change,
        Log2_FC = ifelse(fold_change > 0, log2(fold_change), NA),
        T_Statistic = ttest$statistic,
        P_Value = ttest$p.value,
        Significant = ttest$p.value < 0.05
      ))
    }
  }
  
  diff_results <- diff_results[order(diff_results$P_Value),]
  write.csv(diff_results, "differential_analysis_results.csv", row.names = FALSE)
  
  # Volcano plot
  png("volcano_plot.png", width = 1000, height = 800)
  plot_data <- diff_results[!is.na(diff_results$Log2_FC) & !is.na(diff_results$P_Value),]
  
  colors <- ifelse(plot_data$P_Value < 0.05 & abs(plot_data$Log2_FC) > 1, "red",
                  ifelse(plot_data$P_Value < 0.05, "orange", "gray"))
  
  plot(plot_data$Log2_FC, -log10(plot_data$P_Value), 
       col = colors, pch = 19, 
       xlab = "Log2 Fold Change (Fasting/Normal)",
       ylab = "-Log10 P-Value",
       main = "Volcano Plot: Normal vs Fasting")
  
  abline(h = -log10(0.05), lty = 2, col = "black")
  abline(v = c(-1, 1), lty = 2, col = "black") 
  grid(TRUE)
  
  dev.off()
  
  cat("Differential analysis completed.\n")
  cat("Significant metabolites (p<0.05):", sum(diff_results$Significant), "\n")
}

# Metabolite concentration heatmap
cat("Creating metabolite concentration heatmap...\n")
top_var_data <- numeric_data_clean[, top_var_metabolites]

# Log transform for better visualization
log_data <- log2(top_var_data + 1e-6)

png("metabolite_heatmap.png", width = 1500, height = 800)
pheatmap(log_data,
         scale = "row",
         color = colorRampPalette(c("blue", "white", "yellow"))(100),
         main = "Metabolite Concentration Heatmap (Top 50 Most Variable)",
         fontsize = 8,
         fontsize_row = 8,
         fontsize_col = 6)
dev.off()
cat("Metabolite heatmap saved to metabolite_heatmap.png\n")

# Generate comprehensive report
cat("Generating analysis report...\n")

report_lines <- c(
  "# Metabolomics Analysis Report",
  "",
  paste("**Analysis Date:**", Sys.Date()),
  "**Data Source:** fasting.csv",
  "",
  "## Data Overview",
  paste("- **Dimensions:**", nrow(numeric_data), "samples ×", ncol(numeric_data), "metabolites"),
  "- **Sample Types:** Normal control and 12h fasting conditions", 
  "- **Data Type:** Metabolite concentration measurements",
  "",
  "## Data Quality Assessment",
  paste("- **Missing values:**", sum(is.na(numeric_data)), "/", length(numeric_data), 
        paste0("(", round(sum(is.na(numeric_data))/length(numeric_data)*100, 2), "%)")),
  paste("- **Data completeness:**", round(100 - sum(is.na(numeric_data))/length(numeric_data)*100, 2), "%"),
  "",
  "## Statistical Summary", 
  paste("- **Concentration range:**", sprintf("%.6f", min(summary_stats$Mean)), "to", sprintf("%.6f", max(summary_stats$Mean))),
  paste("- **Most variable metabolites:**", paste(head(summary_stats$Metabolite[order(-summary_stats$SD)], 3), collapse = ", ")),
  "",
  "## Principal Component Analysis Results",
  paste("- **PC1 variance explained:**", paste0(round(summary(pca_result)$importance[2,1]*100, 1), "%")),
  paste("- **PC2 variance explained:**", paste0(round(summary(pca_result)$importance[2,2]*100, 1), "%")),
  paste("- **Total variance explained (PC1+PC2):**", paste0(round(sum(summary(pca_result)$importance[2,1:2])*100, 1), "%")),
  "",
  "## Correlation Analysis Results",
  paste("- **High correlations (>0.7):**", nrow(high_corr_pairs), "metabolite pairs identified"),
  ifelse(nrow(high_corr_pairs) > 0, 
         paste("- **Strongest correlation:**", sprintf("%.3f", high_corr_pairs$Correlation[1]), 
               "between", substr(high_corr_pairs$Metabolite1[1], 1, 30), "and", substr(high_corr_pairs$Metabolite2[1], 1, 30)),
         ""),
  ""
)

if(exists("diff_results") && nrow(diff_results) > 0) {
  diff_lines <- c(
    "## Differential Analysis Results (Normal vs Fasting)",
    paste("- **Total metabolites analyzed:**", nrow(diff_results)),
    paste("- **Significantly different metabolites (p<0.05):**", sum(diff_results$Significant, na.rm = TRUE)),
    paste("- **Upregulated in fasting:**", sum(diff_results$Significant & diff_results$Log2_FC > 0, na.rm = TRUE)),
    paste("- **Downregulated in fasting:**", sum(diff_results$Significant & diff_results$Log2_FC < 0, na.rm = TRUE)),
    paste("- **Most significant metabolite:**", substr(diff_results$Metabolite[1], 1, 50), 
          paste0("(p=", sprintf("%.2e", diff_results$P_Value[1]), ")")),
    ""
  )
  report_lines <- c(report_lines, diff_lines)
}

final_lines <- c(
  "## Generated Files",
  "- `pca_analysis.png` - Principal Component Analysis plots",
  "- `correlation_heatmap.png` - Metabolite correlation visualization",
  "- `metabolite_heatmap.png` - Concentration heatmap of top variable metabolites",
  "- `volcano_plot.png` - Differential analysis volcano plot",
  "- `summary_statistics.csv` - Statistical summary of all metabolites", 
  "- `high_correlations.csv` - List of highly correlated metabolite pairs",
  "- `differential_analysis_results.csv` - Complete differential analysis results",
  "- `analysis_report.md` - This comprehensive report",
  "",
  "## Interpretation & Insights",
  "",
  "### PCA Insights",
  "The PCA analysis reveals the main sources of metabolic variation between normal and fasting states.",
  "The first two principal components capture the primary metabolic differences, helping identify", 
  "which metabolites contribute most to the distinction between conditions.",
  "",
  "### Correlation Patterns", 
  "The correlation analysis identifies metabolites that show coordinated changes, potentially indicating:",
  "- Shared metabolic pathways",
  "- Co-regulated metabolites", 
  "- Technical correlations in measurement",
  "",
  "### Differential Analysis",
  "The comparison between normal and fasting states reveals metabolites that show significant changes",
  "during the fasting period. These could represent:",
  "- Metabolic adaptations to fasting",
  "- Biomarkers of fasting state", 
  "- Key metabolites in energy metabolism",
  "",
  "## Conclusion", 
  "This comprehensive metabolomics analysis provides insights into the metabolic changes associated",
  "with 12-hour fasting. The results identify key metabolites and patterns that distinguish fasting",
  "from normal metabolic states, which may be valuable for understanding metabolic regulation and",
  "identifying potential biomarkers.",
  "",
  "---",
  "*Analysis performed with R-based metabolomics workflow*"
)

report_lines <- c(report_lines, final_lines)

writeLines(report_lines, "analysis_report.md")

cat("\n=== Analysis Complete! ===\n")
cat("All results saved to current directory.\n")
cat("Generated files:\n")
cat("- pca_analysis.png\n")
cat("- correlation_heatmap.png\n") 
cat("- metabolite_heatmap.png\n")
cat("- volcano_plot.png\n")
cat("- summary_statistics.csv\n")
cat("- high_correlations.csv\n")
cat("- differential_analysis_results.csv\n")
cat("- analysis_report.md\n")

File: metabolomics_analysis.py
#!/usr/bin/env python3
"""
Comprehensive Metabolomics Analysis Workflow
Analysis Date: 2025-09-05
Data Source: fasting.csv
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from scipy import stats
import warnings
warnings.filterwarnings('ignore')

def load_data(filename):
    """Load and preprocess metabolomics data"""
    print("Loading metabolomics data...")
    try:
        # Try reading with different separators
        data = pd.read_csv(filename, sep=';', index_col=0)
        print(f"Data loaded successfully with semicolon separator")
    except:
        try:
            data = pd.read_csv(filename, sep=',', index_col=0)
            print(f"Data loaded successfully with comma separator")
        except Exception as e:
            print(f"Error loading data: {e}")
            return None
    
    print(f"Data dimensions: {data.shape}")
    print(f"Columns: {data.columns.tolist()[:10]}...")  # Show first 10 columns
    return data

def preprocess_data(data):
    """Clean and preprocess the data"""
    print("\nPreprocessing data...")
    
    # Convert to numeric, handling any text values
    numeric_data = data.apply(pd.to_numeric, errors='coerce')
    
    # Remove columns with all NaN values
    numeric_data = numeric_data.dropna(axis=1, how='all')
    
    # Basic data info
    print(f"Numeric data shape: {numeric_data.shape}")
    missing_values = numeric_data.isnull().sum().sum()
    total_values = numeric_data.size
    missing_percent = (missing_values / total_values) * 100
    print(f"Missing values: {missing_values}/{total_values} ({missing_percent:.2f}%)")
    
    return numeric_data

def basic_statistics(data):
    """Calculate basic statistics"""
    print("\nCalculating basic statistics...")
    
    stats_summary = pd.DataFrame({
        'Mean': data.mean(),
        'Std': data.std(),
        'Min': data.min(),
        'Max': data.max(),
        'Median': data.median(),
        'Q25': data.quantile(0.25),
        'Q75': data.quantile(0.75)
    })
    
    # Save summary statistics
    stats_summary.to_csv('summary_statistics.csv')
    print("Basic statistics saved to summary_statistics.csv")
    
    return stats_summary

def perform_pca_analysis(data):
    """Perform Principal Component Analysis"""
    print("\nPerforming PCA analysis...")
    
    # Standardize the data
    scaler = StandardScaler()
    data_scaled = scaler.fit_transform(data.fillna(0))
    
    # Perform PCA
    pca = PCA()
    pca_result = pca.fit_transform(data_scaled)
    
    # Create PCA dataframe
    pca_df = pd.DataFrame(pca_result, 
                         columns=[f'PC{i+1}' for i in range(pca_result.shape[1])],
                         index=data.index)
    
    # Plot PCA
    plt.figure(figsize=(12, 8))
    
    # Create subplot for PCA plot
    plt.subplot(2, 2, 1)
    colors = ['red' if 'normal' in idx.lower() else 'blue' for idx in data.index]
    plt.scatter(pca_df['PC1'], pca_df['PC2'], c=colors, alpha=0.7)
    plt.xlabel(f'PC1 ({pca.explained_variance_ratio_[0]:.1%} variance)')
    plt.ylabel(f'PC2 ({pca.explained_variance_ratio_[1]:.1%} variance)')
    plt.title('PCA Analysis - Sample Distribution')
    plt.grid(True, alpha=0.3)
    
    # Add sample labels
    for i, idx in enumerate(data.index):
        plt.annotate(idx, (pca_df.iloc[i]['PC1'], pca_df.iloc[i]['PC2']), 
                    fontsize=8, alpha=0.7)
    
    # Variance explained plot
    plt.subplot(2, 2, 2)
    cumvar = np.cumsum(pca.explained_variance_ratio_)[:10]
    plt.plot(range(1, len(cumvar)+1), cumvar, 'bo-')
    plt.xlabel('Principal Component')
    plt.ylabel('Cumulative Variance Explained')
    plt.title('PCA Variance Explained')
    plt.grid(True, alpha=0.3)
    
    # Feature contributions to PC1
    plt.subplot(2, 2, 3)
    pc1_contributions = pd.DataFrame({
        'Metabolite': data.columns,
        'PC1_Loading': pca.components_[0]
    }).sort_values('PC1_Loading', key=abs, ascending=False).head(20)
    
    plt.barh(range(len(pc1_contributions)), pc1_contributions['PC1_Loading'])
    plt.yticks(range(len(pc1_contributions)), 
               [m[:20]+'...' if len(m)>20 else m for m in pc1_contributions['Metabolite']])
    plt.xlabel('PC1 Loading')
    plt.title('Top 20 Metabolite Contributions to PC1')
    plt.grid(True, alpha=0.3)
    
    # Feature contributions to PC2
    plt.subplot(2, 2, 4)
    pc2_contributions = pd.DataFrame({
        'Metabolite': data.columns,
        'PC2_Loading': pca.components_[1]
    }).sort_values('PC2_Loading', key=abs, ascending=False).head(20)
    
    plt.barh(range(len(pc2_contributions)), pc2_contributions['PC2_Loading'])
    plt.yticks(range(len(pc2_contributions)), 
               [m[:20]+'...' if len(m)>20 else m for m in pc2_contributions['Metabolite']])
    plt.xlabel('PC2 Loading')
    plt.title('Top 20 Metabolite Contributions to PC2')
    plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig('pca_analysis.png', dpi=300, bbox_inches='tight')
    print("PCA analysis saved to pca_analysis.png")
    
    return pca, pca_df

def correlation_analysis(data):
    """Perform correlation analysis"""
    print("\nPerforming correlation analysis...")
    
    # Calculate correlation matrix
    corr_matrix = data.corr()
    
    # Find high correlations
    high_corr_pairs = []
    for i in range(len(corr_matrix.columns)):
        for j in range(i+1, len(corr_matrix.columns)):
            corr_val = corr_matrix.iloc[i, j]
            if abs(corr_val) > 0.7:
                high_corr_pairs.append({
                    'Metabolite1': corr_matrix.columns[i],
                    'Metabolite2': corr_matrix.columns[j],
                    'Correlation': corr_val
                })
    
    high_corr_df = pd.DataFrame(high_corr_pairs).sort_values('Correlation', 
                                                           key=abs, 
                                                           ascending=False)
    high_corr_df.to_csv('high_correlations.csv', index=False)
    print(f"Found {len(high_corr_pairs)} high correlation pairs (>0.7)")
    
    # Create correlation heatmap for top metabolites
    plt.figure(figsize=(15, 12))
    
    # Select top 50 most variable metabolites for visualization
    top_metabolites = data.std().nlargest(50).index
    corr_subset = corr_matrix.loc[top_metabolites, top_metabolites]
    
    sns.heatmap(corr_subset, 
                cmap='RdBu_r', 
                center=0, 
                square=True,
                fmt='.2f',
                cbar_kws={'label': 'Correlation'})
    plt.title('Metabolite Correlation Matrix (Top 50 Most Variable)')
    plt.xticks(rotation=45, ha='right')
    plt.yticks(rotation=0)
    plt.tight_layout()
    plt.savefig('correlation_heatmap.png', dpi=300, bbox_inches='tight')
    print("Correlation heatmap saved to correlation_heatmap.png")
    
    return corr_matrix, high_corr_df

def differential_analysis(data):
    """Compare normal vs fasting conditions"""
    print("\nPerforming differential analysis...")
    
    # Separate normal and fasting samples
    normal_samples = [idx for idx in data.index if 'normal' in idx.lower()]
    fasting_samples = [idx for idx in data.index if 'fasting' in idx.lower()]
    
    print(f"Normal samples: {len(normal_samples)}")
    print(f"Fasting samples: {len(fasting_samples)}")
    
    if len(normal_samples) > 0 and len(fasting_samples) > 0:
        # Perform t-tests for each metabolite
        ttest_results = []
        
        for metabolite in data.columns:
            normal_values = data.loc[normal_samples, metabolite].dropna()
            fasting_values = data.loc[fasting_samples, metabolite].dropna()
            
            if len(normal_values) >= 2 and len(fasting_values) >= 2:
                # Perform t-test
                stat, pvalue = stats.ttest_ind(normal_values, fasting_values)
                
                # Calculate fold change
                normal_mean = normal_values.mean()
                fasting_mean = fasting_values.mean()
                if normal_mean > 0:
                    fold_change = fasting_mean / normal_mean
                else:
                    fold_change = np.nan
                
                ttest_results.append({
                    'Metabolite': metabolite,
                    'Normal_Mean': normal_mean,
                    'Fasting_Mean': fasting_mean,
                    'Fold_Change': fold_change,
                    'Log2_FC': np.log2(fold_change) if fold_change > 0 else np.nan,
                    'T_Statistic': stat,
                    'P_Value': pvalue,
                    'Significant': pvalue < 0.05
                })
        
        # Convert to dataframe and sort by p-value
        ttest_df = pd.DataFrame(ttest_results).sort_values('P_Value')
        ttest_df.to_csv('differential_analysis_results.csv', index=False)
        
        # Create volcano plot
        plt.figure(figsize=(10, 8))
        
        # Filter out invalid values for plotting
        plot_data = ttest_df.dropna(subset=['Log2_FC', 'P_Value'])
        log_pval = -np.log10(plot_data['P_Value'])
        
        # Color points based on significance and fold change
        colors = []
        for _, row in plot_data.iterrows():
            if row['P_Value'] < 0.05 and abs(row['Log2_FC']) > 1:
                colors.append('red')  # Significant and large fold change
            elif row['P_Value'] < 0.05:
                colors.append('orange')  # Significant but small fold change
            else:
                colors.append('gray')  # Not significant
        
        plt.scatter(plot_data['Log2_FC'], log_pval, c=colors, alpha=0.6)
        plt.xlabel('Log2 Fold Change (Fasting/Normal)')
        plt.ylabel('-Log10 P-Value')
        plt.title('Volcano Plot: Normal vs Fasting')
        
        # Add significance lines
        plt.axhline(y=-np.log10(0.05), color='black', linestyle='--', alpha=0.5)
        plt.axvline(x=1, color='black', linestyle='--', alpha=0.5)
        plt.axvline(x=-1, color='black', linestyle='--', alpha=0.5)
        
        plt.grid(True, alpha=0.3)
        plt.tight_layout()
        plt.savefig('volcano_plot.png', dpi=300, bbox_inches='tight')
        
        print(f"Differential analysis completed. {len(ttest_df)} metabolites analyzed.")
        print(f"Significant metabolites (p<0.05): {sum(ttest_df['Significant'])}")
        
        return ttest_df
    
    return None

def create_metabolite_heatmap(data):
    """Create comprehensive metabolite heatmap"""
    print("\nCreating metabolite concentration heatmap...")
    
    # Select top most variable metabolites
    top_metabolites = data.std().nlargest(50).index
    data_subset = data.loc[:, top_metabolites]
    
    # Create heatmap
    plt.figure(figsize=(20, 10))
    
    # Log transform for better visualization (add small constant to avoid log(0))
    data_log = np.log2(data_subset + 1e-6)
    
    sns.heatmap(data_log, 
                cmap='viridis',
                cbar_kws={'label': 'Log2 Concentration'},
                xticklabels=[col[:30]+'...' if len(col)>30 else col for col in data_subset.columns],
                yticklabels=data_subset.index)
    
    plt.title('Metabolite Concentration Heatmap (Top 50 Most Variable)')
    plt.xlabel('Metabolites')
    plt.ylabel('Samples')
    plt.xticks(rotation=45, ha='right')
    plt.tight_layout()
    plt.savefig('metabolite_heatmap.png', dpi=300, bbox_inches='tight')
    print("Metabolite heatmap saved to metabolite_heatmap.png")

def generate_report(data, stats_summary, pca, high_corr_df, ttest_df=None):
    """Generate comprehensive analysis report"""
    print("\nGenerating analysis report...")
    
    report = f"""# Metabolomics Analysis Report

**Analysis Date:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}
**Data Source:** fasting.csv

## Data Overview
- **Dimensions:** {data.shape[0]} samples × {data.shape[1]} metabolites
- **Sample Types:** Normal control and 12h fasting conditions
- **Data Type:** Metabolite concentration measurements

## Data Quality Assessment
- **Missing values:** {data.isnull().sum().sum()}/{data.size} ({(data.isnull().sum().sum()/data.size)*100:.2f}%)
- **Data completeness:** {100-(data.isnull().sum().sum()/data.size)*100:.2f}%

## Statistical Summary
- **Concentration range:** {stats_summary['Mean'].min():.6f} to {stats_summary['Mean'].max():.6f}
- **Most variable metabolites:** {', '.join(data.std().nlargest(5).index[:3])}...

## Principal Component Analysis Results
- **PC1 variance explained:** {pca.explained_variance_ratio_[0]:.1%}
- **PC2 variance explained:** {pca.explained_variance_ratio_[1]:.1%}
- **Total variance explained (PC1+PC2):** {(pca.explained_variance_ratio_[0] + pca.explained_variance_ratio_[1]):.1%}

## Correlation Analysis Results
- **High correlations (>0.7):** {len(high_corr_df)} metabolite pairs identified
- **Strongest positive correlation:** {high_corr_df.iloc[0]['Correlation']:.3f} between {high_corr_df.iloc[0]['Metabolite1'][:30]}... and {high_corr_df.iloc[0]['Metabolite2'][:30]}...

"""

    if ttest_df is not None:
        significant_metabolites = sum(ttest_df['Significant'])
        upregulated = sum((ttest_df['Significant']) & (ttest_df['Log2_FC'] > 0))
        downregulated = sum((ttest_df['Significant']) & (ttest_df['Log2_FC'] < 0))
        
        report += f"""## Differential Analysis Results (Normal vs Fasting)
- **Total metabolites analyzed:** {len(ttest_df)}
- **Significantly different metabolites (p<0.05):** {significant_metabolites}
- **Upregulated in fasting:** {upregulated}
- **Downregulated in fasting:** {downregulated}
- **Most significant metabolite:** {ttest_df.iloc[0]['Metabolite'][:50]}... (p={ttest_df.iloc[0]['P_Value']:.2e})

"""

    report += """## Generated Files
- `pca_analysis.png` - Principal Component Analysis plots
- `correlation_heatmap.png` - Metabolite correlation visualization  
- `metabolite_heatmap.png` - Concentration heatmap of top variable metabolites
- `volcano_plot.png` - Differential analysis volcano plot
- `summary_statistics.csv` - Statistical summary of all metabolites
- `high_correlations.csv` - List of highly correlated metabolite pairs
- `differential_analysis_results.csv` - Complete differential analysis results
- `analysis_report.md` - This comprehensive report

## Interpretation & Insights

### PCA Insights
The PCA analysis reveals the main sources of metabolic variation between normal and fasting states. The first two principal components capture the primary metabolic differences, helping identify which metabolites contribute most to the distinction between conditions.

### Correlation Patterns
The correlation analysis identifies metabolites that show coordinated changes, potentially indicating:
- Shared metabolic pathways
- Co-regulated metabolites
- Technical correlations in measurement

### Differential Analysis
The comparison between normal and fasting states reveals metabolites that show significant changes during the fasting period. These could represent:
- Metabolic adaptations to fasting
- Biomarkers of fasting state
- Key metabolites in energy metabolism

## Conclusion
This comprehensive metabolomics analysis provides insights into the metabolic changes associated with 12-hour fasting. The results identify key metabolites and patterns that distinguish fasting from normal metabolic states, which may be valuable for understanding metabolic regulation and identifying potential biomarkers.

---
*Analysis performed with Python-based metabolomics workflow*
"""

    with open('analysis_report.md', 'w') as f:
        f.write(report)
    
    print("Analysis report saved to analysis_report.md")

def main():
    """Main analysis workflow"""
    print("=== Comprehensive Metabolomics Analysis Workflow ===")
    print("Starting analysis...")
    
    # Load data
    data = load_data('fasting.csv')
    if data is None:
        print("Failed to load data. Exiting.")
        return
    
    # Preprocess data
    numeric_data = preprocess_data(data)
    
    # Basic statistics
    stats_summary = basic_statistics(numeric_data)
    
    # PCA analysis
    pca, pca_df = perform_pca_analysis(numeric_data)
    
    # Correlation analysis
    corr_matrix, high_corr_df = correlation_analysis(numeric_data)
    
    # Differential analysis
    ttest_df = differential_analysis(numeric_data)
    
    # Create heatmap
    create_metabolite_heatmap(numeric_data)
    
    # Generate comprehensive report
    generate_report(numeric_data, stats_summary, pca, high_corr_df, ttest_df)
    
    print("\n=== Analysis Complete! ===")
    print("All results saved to current directory.")

if __name__ == "__main__":
    main()